### PR TITLE
add `redis_json_dict` for persistent metadata

### PR DESCRIPTION
**Pull request by hyperrealist**
Uses https://pypi.org/project/redis-json-dict.

### CODE REVIEW

**Review by mrakitin — APPROVED** [review]
The import step passed in CI, there is another unrelated failure:
```py
Executing /home/vsts/work/1/s/startup/00-base.py in CI
Loading /home/vsts/work/1/s/startup/00-base.py...

EpicsSignalBase timeout is 10 [seconds]. NOT going to touch beamline hardware.

OBJECT CACHE: Will use up to 1_088_732_774 bytes (15% of total physical RAM)
INFO:tiled.server.object_cache:Will use up to 1_088_732_774 bytes (15% of total physical RAM)
Cannot install event loop hook for "qt5" when running with `--simple-prompt`.
NOTE: Tk is supported natively; use Tk apps and Tk backends with `--simple-prompt`.
environment variable BLUESKY_LOG_FILE is not set, using default log file path '/home/vsts/.cache/bluesky/log/bluesky.log'
environment variable BLUESKY_IPYTHON_LOG_FILE is not set, using default file path '/home/vsts/.cache/bluesky/log/bluesky_ipython.log'
Activating auto-logging. Current session state plus future input saved.
Filename       : /home/vsts/.cache/bluesky/log/bluesky_ipython.log
Mode           : append
Output logging : True
Raw input log  : False
Timestamping   : True
State          : active
%3|1708030184.645|FAIL|rdkafka#producer-1| [thrd:localhost:9092/bootstrap]: localhost:9092/bootstrap: Connect to ipv4#127.0.0.1:9092 failed: Connection refused (after 0ms in state CONNECT)
%3|1708030185.644|FAIL|rdkafka#producer-1| [thrd:localhost:9092/bootstrap]: localhost:9092/bootstrap: Connect to ipv6#[::1]:9092 failed: Connection refused (after 0ms in state CONNECT)
%3|1708030186.644|FAIL|rdkafka#producer-1| [thrd:localhost:9092/bootstrap]: localhost:9092/bootstrap: Connect to ipv4#127.0.0.1:9092 failed: Connection refused (after 0ms in state CONNECT)
%3|1708030187.644|FAIL|rdkafka#producer-1| [thrd:localhost:9092/bootstrap]: localhost:9092/bootstrap: Connect to ipv4#127.0.0.1:9092 failed: Connection refused (after 0ms in state CONNECT, 1 identical error(s) suppressed)
%3|1708030188.644|FAIL|rdkafka#producer-1| [thrd:localhost:9092/bootstrap]: localhost:9092/bootstrap: Connect to ipv6#[::1]:9092 failed: Connection refused (after 0ms in state CONNECT)
%3|1708030190.644|FAIL|rdkafka#producer-1| [thrd:localhost:9092/bootstrap]: localhost:9092/bootstrap: Connect to ipv4#127.0.0.1:9092 failed: Connection refused (after 0ms in state CONNECT)
%3|1708030191.644|FAIL|rdkafka#producer-1| [thrd:localhost:9092/bootstrap]: localhost:9092/bootstrap: Connect to ipv6#[::1]:9092 failed: Connection refused (after 0ms in state CONNECT)
%3|1708030193.644|FAIL|rdkafka#producer-1| [thrd:localhost:9092/bootstrap]: localhost:9092/bootstrap: Connect to ipv4#127.0.0.1:9092 failed: Connection refused (after 0ms in state CONNECT)
Exception reporting mode: Minimal
OBJECT CACHE: Will use up to 1_088_732_774 bytes (15% of total physical RAM)
INFO:tiled.server.object_cache:Will use up to 1_088_732_774 bytes (15% of total physical RAM)
An exception has occurred, use '%tb verbose' to see the full traceback.
KeyError: 'TILED_BLUESKY_WRITING_API_KEY_SRX'
```

**Comment by hyperrealist** [comment]
I remember seeing that error somewhere before

File: startup/02-usersetup.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+# proposal_num = 313507
+# proposal_title = 'Data security at SRX'
+# PI_lastname = 'Kiss'
+# saf_num = 312779
+
+
+cycle = "2024_cycle1"
+
+# Set user data in bluesky
+# RE.md['data_session'] = data_session
```
Comment (mrakitin) [inline startup/02-usersetup.py:54]:
```suggestion
```

File: startup/31-xspress3.py [hunk 2]
```diff
@@ -210,6 +210,8 @@ def root_path_str(self):
         # cycle = self._redis_dict["cycle"]
         data_session = RE.md["data_session"]
         cycle = RE.md["proposal"]["cycle"]
+        print(f"{data_session=}")
+        print(f"{cycle=}")
```
Comment (mrakitin) [inline startup/31-xspress3.py:214]: Those lines should be deleted after testing.
```suggestion
```

File: startup/02-usersetup.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import time as ttime
+
+import httpx
+
+### Proposal information put into the metadata
+# proposal_num = None
+# proposal_title = None
+# PI_lastname = None
+# saf_num = None
+
```
Comment (mrakitin) [inline startup/02-usersetup.py:14]:
```suggestion
```

File: startup/00-base.py [hunk 4]
```diff
@@ -113,112 +117,31 @@ def post_document(name, doc):
...
+                Token.Prompt,
+                f"☁️  BlueSky@SRX | Proposal #{RE.md.get('proposal', {}).get('proposal_num', 'N/A')} [",
+            ),
+            (Token.PromptNum, str(self.shell.execution_count)),
+            (Token.Prompt, "]: "),
+        ]
+
+
+ip.prompts = SRXPrompt(ip)
+""
```
Comment (mrakitin) [inline startup/00-base.py:140]:
```suggestion

```

**Review by mrakitin — COMMENTED** [review]
Some suggested changes are below.

File: startup/00-base.py (hunk 4 above)
Comment (hyperrealist) [reply startup/00-base.py:140]: Not a huge deal, but I found this strange indentation behavior when you hit return. Perhaps limited to my terminal.

image

Comment (mrakitin) [reply startup/00-base.py:140]: We can remove the ☁️ symbol, it was more like a joke.

File: startup/00-base.py [hunk 5]
```diff
@@ -54,53 +68,43 @@ def wait_for_connection(self, timeout=DEFAULT_CONNECTION_TIMEOUT):
...
-print(f'\nEpicsSignalBase timeout is {timeout} [seconds]. {going} to touch beamline hardware.\n')
+print(f"\nEpicsSignalBase timeout is {timeout} [seconds]. {going} to touch beamline hardware.\n")
 
-from ophyd.signal import EpicsSignalBase
 # EpicsSignalBase.set_default_timeout(timeout=timeout, connection_timeout=timeout)  # old style
 EpicsSignalBase.set_defaults(timeout=timeout, connection_timeout=timeout)  # new style
 
-import datetime
 
 def print_now():
```
Comment (mrakitin) [inline startup/00-base.py:77]: Looks like we have this `print_now` function defined twice. Need to keep just once copy.

**Comment by andrewmkiss** [comment]
@mrakitin @hyperrealist Since this is the latest branch I think it is a good idea to merge with `working`. I would like to play around during my time this week but by Friday i am hoping to pull `working` into `master`. Does that seem reasonable? 
Can/should I merge this PR? Otherwise, I looked at the changes and I approve the PR if someone else wants to do it

**Comment by mrakitin** [comment]
@andrewmkiss, I like the idea! Do you want me to do some cleanup indicated above first?

**Comment by hyperrealist** [comment]
> @mrakitin @hyperrealist Since this is the latest branch I think it is a good idea to merge with `working`. I would like to play around during my time this week but by Friday i am hoping to pull `working` into `master`. Does that seem reasonable? 
> 
> Can/should I merge this PR? Otherwise, I looked at the changes and I approve the PR if someone else wants to do it

@andrewmkiss I somehow did not see this message timely in my notifications feed. Sorry about that, and thank you for merging!

**Pull request merged**